src/AutoRest.CSharp/Mgmt/Generation/ResourceWriter.cs

```diff
@@ -1,37 +1,61 @@
 using Azure.ResourceManager.Core;
+using AutoRest.CSharp.Mgmt.AutoRest;
+using AutoRest.CSharp.Output.Models.Types;
 
 namespace AutoRest.CSharp.Mgmt.Generation
 {
     internal class ResourceWriter
     {
-        public void WriteResource(CodeWriter writer, BuildContext<MgmtOutputLibrary> context, Resource resource)
+        public void WriteResource(CodeWriter writer, Resource resource, BuildContext<MgmtOutputLibrary> context)
         {
-            using (writer.Namespace(resource.Type.Namespace))
+            var cs = resource.Type;
+
+            using (writer.Namespace(cs.Namespace))
             {
-                var resourceOperation = context.Library.GetResourceOperation(resource.OperationGroup);
                 writer.WriteXmlDocumentationSummary(resource.Description);
-                using (writer.Scope($"{resource.Declaration.Accessibility} class {resource.Type.Name} : {resourceOperation.Type}"))
+
+                using (writer.Scope($"{resource.Declaration.Accessibility} class {cs.Name} : {cs.Name}Operations"))
                 {
-                    var resourceData = context.Library.GetResourceData(resource.OperationGroup);
-                    WriteConstructors(writer, resource, resourceData);
-                }
-            }
-        }
+                    // internal constructor
+                    writer.WriteXmlDocumentationSummary($"Initializes a new instance of the <see cref = \"{cs.Name}\"/> class.");
+                    writer.WriteXmlDocumentationParameter("options", "The client parameters to use in these operations.");
+                    writer.WriteXmlDocumentationParameter("resource", "The resource that is the target of operations.");
+                    using (writer.Scope($"internal {cs.Name}({typeof(ResourceOperationsBase)} options, {context.Library.GetResourceData(resource.OperationGroup).Type} resource) : base(options, resource.Id)"))
+                    {
+                        writer.LineRaw("Data = resource;");
+                    }
 
-        private void WriteConstructors(CodeWriter writer, Resource resource, ResourceData resourceData)
-        {
-            writer.WriteXmlDocumentationSummary($"Initializes a new instance of the <see cref=\"{resource.Type.Name}\"/> class.");
-            writer.WriteXmlDocumentationParameter("options", "The client parameters to use in these operations.");
-            writer.WriteXmlDocumentationParameter("resource", "The resource that is the target of operations.");
-            using (writer.Scope($"internal {resource.Type.Name}({typeof(ResourceOperationsBase)} options, {resourceData.Type} resource) : base()"))
-            {
+                    // write Data
+                    writer.Line();
+                    writer.WriteXmlDocumentationSummary($"Gets or sets the {context.Library.GetResourceData(resource.OperationGroup).Type.Name}.");
+                    var resourceDataObject = context.Library.GetResourceData(resource.OperationGroup);
```

Comment:
`GetResourceData(operationGroup)` is called multiple times. I would suggest to move `resourceDataObject` variable before `using` and use it everywhere, instead of calling `GetResourceData` multiple times.